File: app/src/main/java/com/june/studyproject/service/TestBindService.java
package com.june.studyproject.service;

import android.app.Service;
import android.content.Intent;
import android.os.Binder;
import android.os.IBinder;
import android.util.Log;

import androidx.annotation.Nullable;

import com.june.studyproject.R;

import java.util.Calendar;

public class TestBindService extends Service {

    private static final String TAG = "TestBindService";

    @Override
    public void onCreate() {
        super.onCreate();
        Log.e(TAG, "onCreate: 创建TestBindService");
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.e(TAG, "onStartCommand: 启动TestBindService");
        return super.onStartCommand(intent, flags, startId);
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        Log.e(TAG, "onBind: 绑定TestBindService");
        return new TestBinder();
    }

    @Override
    public void onRebind(Intent intent) {
        super.onRebind(intent);
        Log.e(TAG, "onRebind: 再次绑定TestBindService");
    }

    @Override
    public boolean onUnbind(Intent intent) {
        Log.e(TAG, "onUnbind: 解绑TestBindService");
        //return super.onUnbind(intent);
        return true;
    }

    @Override
    public boolean stopService(Intent name) {
        Log.e(TAG, "stopService: 关闭TestBindService");
        return super.stopService(name);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.e(TAG, "onDestroy: 销毁TestBindService");
    }

    public String getServiceInformation() {
        Calendar calendar = Calendar.getInstance();

        return getString(R.string.handle_service_time,
                calendar.get(Calendar.YEAR),
                calendar.get(Calendar.MONTH) + 1,
                calendar.get(Calendar.DAY_OF_MONTH),
                calendar.get(Calendar.HOUR),
                calendar.get(Calendar.SECOND));
    }

    public class TestBinder extends Binder {
        public TestBindService getService() {
            return TestBindService.this;
        }
    }
}

File: app/src/main/java/com/june/studyproject/component/service/TestBindService.kt
package com.june.studyproject.component.service

import android.app.Service
import android.content.Intent
import android.os.Binder
import android.os.IBinder
import com.june.studyproject.R
import timber.log.Timber
import java.util.*

class TestBindService : Service() {
    override fun onCreate() {
        super.onCreate()
        Timber.e("onCreate: 创建TestBindService")
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        Timber.e("onStartCommand: 启动TestBindService")
        return super.onStartCommand(intent, flags, startId)
    }

    override fun onBind(intent: Intent): IBinder? {
        Timber.e("onBind: 绑定TestBindService")
        return TestBinder()
    }

    override fun onRebind(intent: Intent) {
        super.onRebind(intent)
        Timber.e("onRebind: 再次绑定TestBindService")
    }

    override fun onUnbind(intent: Intent): Boolean {
        Timber.e("onUnbind: 解绑TestBindService")
        //return super.onUnbind(intent);
        return true
    }

    override fun stopService(name: Intent): Boolean {
        Timber.e("stopService: 关闭TestBindService")
        return super.stopService(name)
    }

    override fun onDestroy() {
        super.onDestroy()
        Timber.e("onDestroy: 销毁TestBindService")
    }

    val serviceInformation: String
        get() {
            val calendar = Calendar.getInstance()
            return getString(R.string.handle_service_time,
                             calendar[Calendar.YEAR],
                             calendar[Calendar.MONTH] + 1,
                             calendar[Calendar.DAY_OF_MONTH],
                             calendar[Calendar.HOUR],
                             calendar[Calendar.SECOND])
        }

    inner class TestBinder : Binder() {
        val service: TestBindService
            get() = this@TestBindService
    }
}